### PR TITLE
NMS-16210: Structured Node List documentation

### PR DESCRIPTION
**Pull request by synqotik**
Documentation for the Horizon 33 Structured Node List.

This is a first draft, I expect there will be edits and updates.

### External References

* Jira (Issue Tracker): https://opennms.atlassian.net/browse/NMS-16210

### CODE REVIEW

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+=== Filtering
+
+Filtering allows you to filter by Category (i.e. Surveillance Category), Flows and Monitoring Locations.
+
+Click on an item to filter by that item.
+More than one item per section can be selected.
+As items are selected, the number of items selected will display; a check icon will display to clear that section's selections.
+
+Use "Clear All" to clear all filters and search parameters.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:156]:
```suggestion
Use **Clear All** to clear all filters and search parameters.
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+The Categories section allows you to enable "Match All", in which case only nodes corresponding to *all* the selected categories (i.e. the "intersection") will be displayed.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:158]:
```suggestion
Use **Match All** in the Categories section to display only nodes corresponding to *all* the selected categories.
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+| Search by IP address. Will only search on valid IPv4 and IPv6 addresses, will not search on wildcard or partial addresses
+
+| Sys Contact
+| Search by System Contact. Performs a case-sensitive wildcard search.
+
+| Sys Description
+| Search by System Description. Performs a case-sensitive wildcard search.
+
+| Sys Location
+| Search by System Location. Performs a case-sensitive wildcard search.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:193]:
```suggestion
| Performs a case-sensitive wildcard search by system location.
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+| Sys Name
+| Search by System Name. Performs a case-sensitive wildcard search.
+
+| Sys Object Id
+| Search by System Object ID. Performs a case-sensitive wildcard search.
+
+| SNMP Alias
+| Search by SNMP Alias. Performs an exact search.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:202]:
```suggestion
| Performs an exact search by SNMP alias.
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+| SNMP Description
+| Search by SNMP Description. Performs an exact search.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:205]:
```suggestion
| Performs an exact search by SNMP description.
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+| SNMP Index
+| Search by SNMP Index. Performs an exact search.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:208]:
```suggestion
| Performs an exact search by SNMP index.
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+Above the table is a button that allows you to export the current data in either CSV or JSON format.
+
+A gear icon opens the Node Preferences window which allows you to configure which columns are displayed and in which order.
+Saving this configuration will also save the current search and filtering options.
+
+An input field above the table allows you to search by node label.
+
+On the left is the Filtering and Extended Search panel which allows you to filter and search by various parameters.
+
+== Node Table
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:34]:
```suggestion
== Node table
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+The table displays various columns of information about the nodes that match the current search and filter parameters.
+
+The Node Label and ID fields contain links to navigate to the Node Details page.
+
+The IP Address field contains a link to go to the IP Interface page for that interface.
+A tooltip will also display more information about the interface, for example if it is Managed, Primary, Not Eligible, etc.
+
+The Flows field contains arrows denoting whether Ingress and/or Egress flow information is available, and a tooltip provides more detail.
+
+== Action Menu
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:45]:
```suggestion
== Action menu
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+| Navigate to the Scheduled Outage page
+
+| View Topology
+| Navigate to the Topology Map page
+|===
+
+.Node Action Menu
+image::structured-node-list/node-action-menu.png["{page-component-title} the node Action menu" width=400]
+
+== Node Details
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:99]:
```suggestion
== Node details
```

**Review by Bonrob2 — CHANGES_REQUESTED** [review]
It looks like a lot, but mainly stylistic and style guide updates. Thanks for all this work!

File: docs/modules/operation/images/structured-node-list/node-details.png
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/images/structured-node-list/node-details.png:1]: Yes, fixed in screenshot and in code (both in Node Details dialog and in Extended Search dropdown).

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 10]
```diff
@@ -0,0 +1,218 @@
+
+[[structured-node-list]]
+= Structured Node List
+:description: Provides details about the Structured Node List of {page-component-title}.
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:4]: Done

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+
+The {page-component-title} 33 Web UI includes a new feature, the Structured Node List.
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:6]: Done

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+This is a Vue-based page which displays node information and enables the user to perform enhanced data display, searching, filtering, and data export.
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:7]: Done

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+
+Navigate to this page from the main menu `Info > Structured Node List`.
+
+The main display shows a paginated table of nodes corresponding to the current search and filtering parameters.
+Most of the columns can be sorted in ascending or descending order.
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:12]: Done

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+
+You can choose between 10, 20 or 50 entries per page, and can navigate between pages, or go to the start or end of the currently filtered set.
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:14]: Done

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+
+.Structured Node List
+image::structured-node-list/node-structure-overview.png["{page-component-title} web UI showing structured node list"]
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:17]: Done

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+
+The first column of each node has an Action menu where the user can:
+
+- choose various actions, including "Info..." to open a dialog window displaying details about the node
+- navigate to various relevant pages (Alarms, Events, etc.)
+- perform other actions (Rescan, Schedule an Outage, etc.)
+
+Above the table is a button that allows you to export the current data in either CSV or JSON format.
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:25]: Yes, added, and also added one for the gear symbol.

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+== Action Menu
+
+The Action Menu next to each node contains the following:
+
+[options="autowidth"]
+|===
+| Menu Item              | Description
+
+| Info...
+| Display the Node Details dialog
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:54]: Changed to be a list. Also updated some of the names both in this list and in the dropdown and image.

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+
+| Events
+| Navigate to the Events page
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:57]: Fixed all these to be a list

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+== Node Preferences
+
+Clicking the gear icon opens up the Node Preferences dialog.
+
+.Node Preferences
+image::structured-node-list/node-preferences.png["{page-component-title} the Node Preferences dialog" width=400]
+
+Click "Default" to reset all displayed columns to the default list visibility and order.
+
+Click the checkbox next to each column to display or hide that column. Note that changes take effect immediately.
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:122]: It's admittedly a little tricky. Changes take effect immediately in that they are "saved" in the Vue runtime code. But if user hard-refreshed the page those changes would be lost unless they clicked "Save and Close".

There's some explanation below re the Save feature. It saves by using the web browser's "local storage" facility. So preferences are stored for that browser on that machine only. No one else will see the changes, and that user would not see them if they used a different browser even on same machine, or used a different machine.

Saving preferences by OpenNMS user would be more ideal, but more work, and seems we should coordinate a general "User Preferences" system probably using Content Management.

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+
+Click the up and down arrows to reorder the columns.
+
+Click "Save and Close" to save your settings.
+This will also save your current search and filtering configuration.
+
+Click the "X" to close without making changes.
+
+NOTE: Node Preferences are stored in your browser's `Local Storage`.
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:131]: Good idea, done!

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+Note, this is a case-sensitive wildcard search.
+Entering `cal` will match nodes have the label `localhost` as well as `california` but not `California`.
+
+=== Filtering
+
+Filtering allows you to filter by Category (i.e. Surveillance Category), Flows and Monitoring Locations.
+
+Click on an item to filter by that item.
+More than one item per section can be selected.
+As items are selected, the number of items selected will display; a check icon will display to clear that section's selections.
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:154]: This sentence specifically means the number of items selected is shown.

I updated to:

```
As items are selected, the number of items selected for that section will display.
A check symbol lets you clear the selected items for that section.
```

The image makes it a little clearer: items selected are shaded light gray, and the number appears next to the section name, e.g. "Categories (2) (x)"

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+This allows you to search by various additional fields.
+Note that some fields perform wildcard searches and some are exact searches.
+Also note that some fields, for example SNMP fields, are not displayed in the table, but will display in the Node Details dialog.
+
+[options="autowidth"]
+|===
+| Item                   | Description
+
+| Foreign Source
+| Search by Foreign Source only
```
[the remark this answers is not in the document]
Comment (synqotik) [reply docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:175]: fixed all of these

**Comment by synqotik** [comment]
I believe I addressed everything :)

Note, there's a build failure which I don't think has anything to do with this PR, ticket opened: https://opennms.atlassian.net/browse/NMS-16227

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+
+=== Node label search
+
+The input box above the Node Table lets you search by node label.
+This is a case-sensitive wildcard search.
+Entering `cal` will match nodes have the label `localhost` as well as `california`, but not `California`.
+
+=== Filtering
+
+You can filter by Category (for example, Surveillance Category), flows, and monitoring locations.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:115]:
```suggestion
You can filter by category (for example, surveillance category), flows, and monitoring locations.
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+
+Click on an item to filter by that item.
+You can select more than one item per section.
+
+As items are selected, the number of items selected for that section will display.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:120]:
```suggestion
As you choose items, the number of items selected for that section will display.
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+A check symbol lets you clear the selected items for that section.
+
+Use **Clear All** to clear all filters and search parameters.
+
+Use **Match All** in the Categories section to display only nodes corresponding to *all* the selected categories.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:125]:
```suggestion
Use **Match All** in the Categories section to display only nodes corresponding to all the selected categories.
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 26]
```diff
@@ -0,0 +1,188 @@
...
+
+For example, you could filter on nodes that are in both the *Routers* and *Production* categories.
+
+.Category Filter
+image::structured-node-list/category-filter-match-all.png["{page-component-title} Node Category Filter." width=400]
+
+=== Extended search
+
+This allows you to search by various additional fields.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:134]:
```suggestion
This lets you to search by additional fields.
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 27]
```diff
@@ -0,0 +1,188 @@
+
+[[structured-node-list]]
+= Structured Node List
+:description: Learn about the structured node list in OpenNMS Horizon/Meridian, which lets users perform enhanced data display, search, filter, and export node information.
+
+The Horizon 33 Web UI includes a new feature, the Structured Node List.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:6]:
```suggestion

```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 28]
```diff
@@ -0,0 +1,188 @@
...
+The structured node list is a Vue-based page that displays node information and lets the user perform enhanced data display, search, filter, and export data.
+
+The main display shows a paginated table of nodes corresponding to the current search and filtering parameters.
+You can sort most of the columns in ascending or descending order.
+
+You can choose among 10, 20 or 50 entries per page, navigate between pages, or go to the start or end of the currently filtered set.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:12]:
```suggestion
You can choose among 10, 20, or 50 entries per page, navigate among pages, or go to the start or end of the currently filtered set.
```

File: docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc [hunk 29]
```diff
@@ -0,0 +1,188 @@
...
+
+The first column of each node has an Action menu where the user can do the following:
+
+* Choose various actions, including **Info...** to open a dialog that displays details about the node.
+* Navigate to various relevant pages (Alarms, Events, and so on).
+* Perform other actions (rescan, schedule an outage, and so on).
+
+The **Export** symbol image:structured-node-list/node-export-button.png[width=32] lets you export the current data in either CSV or JSON format.
+
+A gear icon image:structured-node-list/node-preferences-button.png[width=32] opens the Node Preferences window that allows you to configure the columns displayed and their order.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/structured-node-list/introduction.adoc:25]:
```suggestion
A *gear* symbol image:structured-node-list/node-preferences-button.png[width=32] opens the Node Preferences window that allows you to configure the columns displayed and their order.
```

**Review by Bonrob2 — CHANGES_REQUESTED** [review]
A few things I missed the first time through. Thanks!

**Review by Bonrob2 — APPROVED** [review]
Docs LGTM! Thanks!

**Comment by synqotik** [comment]
> A few things I missed the first time through. Thanks!

Fixed! Also merged into `develop` to fix a merge conflict.

**Review by Bonrob2 — APPROVED** [review]
Looks even better now! :-)

**Pull request merged**